src/http/minreq_http.rs

```diff
@@ -0,0 +1,254 @@
+//! This module implements the [`crate::client::Transport`] trait using [`minreq`]
+//! as the underlying HTTP transport.
+//!
+//! [minreq]: <https://github.com/neonmoe/minreq>
+
+#[cfg(jsonrpc_fuzz)]
+use std::io::{self, Read, Write};
+#[cfg(jsonrpc_fuzz)]
+use std::sync::Mutex;
+use std::time::Duration;
+use std::{error, fmt};
+
+use crate::client::Transport;
+use crate::{Request, Response};
+
+const DEFAULT_URL: &str = "http://localhost";
+const DEFAULT_PORT: u16 = 8332; // the default RPC port for bitcoind.
+#[cfg(not(jsonrpc_fuzz))]
+const DEFAULT_TIMEOUT_SECONDS: u64 = 15;
+#[cfg(jsonrpc_fuzz)]
+const DEFAULT_TIMEOUT_SECONDS: u64 = 1;
+
+/// An HTTP transport that uses [`minreq`] and is useful for running a bitcoind RPC client.
+#[derive(Clone, Debug)]
+pub struct MinreqHttpTransport {
+    /// URL of the RPC server.
+    url: String,
+    /// timeout only supports second granularity.
+    timeout: Duration,
+    /// The value of the `Authorization` HTTP header, i.e., a base64 encoding of 'user:password'.
+    basic_auth: Option<String>,
+}
+
+impl Default for MinreqHttpTransport {
+    fn default() -> Self {
+        MinreqHttpTransport {
+            url: format!("{}:{}", DEFAULT_URL, DEFAULT_PORT),
+            timeout: Duration::from_secs(DEFAULT_TIMEOUT_SECONDS),
+            basic_auth: None,
+        }
+    }
+}
+
+impl MinreqHttpTransport {
+    /// Constructs a new [`MinreqHttpTransport`] with default parameters.
+    pub fn new() -> Self {
+        MinreqHttpTransport::default()
+    }
+
+    /// Returns a builder for [`MinreqHttpTransport`].
+    pub fn builder() -> Builder {
+        Builder::new()
+    }
+
+    fn request<R>(&self, req: impl serde::Serialize) -> Result<R, Error>
+    where
+        R: for<'a> serde::de::Deserialize<'a>,
+    {
+        let req = match &self.basic_auth {
+            Some(auth) => minreq::Request::new(minreq::Method::Post, &self.url)
+                .with_timeout(self.timeout.as_secs())
+                .with_header("Authorization", auth)
+                .with_json(&req)?,
+            None => minreq::Request::new(minreq::Method::Post, &self.url)
+                .with_timeout(self.timeout.as_secs())
+                .with_json(&req)?,
+        };
+
+        let resp = req.send()?;
+        let json = resp.json()?;
```

Comment:
Shouldn't we check the `status_code` of the `resp` here?
Otherwise calling `json()` on a response that isn't `200` would try to deserialize an empty body and therefore confusingly return a `SerdeJsonError` (https://github.com/neonmoe/minreq/blob/75f28d4eba632e7652b6c15adc746a251c657ad2/src/response.rs#L164-L177) whereas it's not a JSON error but an invalid request.

Reply:
Bitcoin Core does not return an empty body alongside HTTP error codes. It returns a JSON blob with error information.

Reply:
I think we could just try to parse the JSON, and in the case of error, check whether the response was 200. For 200 we should return the JSON error, for non-200 we should return a status error.

Trying to match Core's exact behavior here, which I doubt is a stable part of their API, seems too fragile.

Reply:
Sounds good!